File: org.kevoree.modeling.microframework/src/main/java/org/kevoree/modeling/api/events/ModelEvent.java
package org.kevoree.modeling.api.events;

import org.kevoree.modeling.api.KObject;
import org.kevoree.modeling.api.meta.Meta;
import org.kevoree.modeling.api.KActionType;

public class ModelEvent {
    private KActionType type;
    private Meta meta;
    private Object pastValue;
    private Object newValue;
    private KObject source;

    public ModelEvent(KActionType type, Meta meta, KObject source, Object pastValue, Object newValue) {
        this.type = type;
        this.meta = meta;
        this.source = source;
        this.pastValue = pastValue;
        this.newValue = newValue;
    }

    public KActionType getType() {
        return type;
    }

    public Meta getMeta() {
        return meta;
    }

    public Object getPastValue() {
        return pastValue;
    }

    public Object getNewValue() {
        return newValue;
    }

    public KObject getSource() {
        return source;
    }

    @Override
    public String toString() {
        return "ModelEvent[src:[" + source.now() + "]" + source.uuid() + ", type:" + type + ", meta:" + getMeta() + ", pastValue:" + getPastValue() + ", newValue:" + getNewValue() + "]";
    }
}

File: org.kevoree.modeling.microframework/src/main/java/org/kevoree/modeling/api/events/ModelEvent.kt
package org.kevoree.modeling.api.events

import org.kevoree.modeling.api.util.ActionType
import org.kevoree.modeling.api.util.ElementAttributeType
import org.kevoree.modeling.api.KObject
import org.kevoree.modeling.api.time.TimeAwareKObject

class ModelEvent(val etype: ActionType, val elementAttributeType: ElementAttributeType, val elementAttributeName: String, val value: Any?, val previous_value: Any?, val source: KObject?, val previousPath: String?) {

    override fun toString(): String {
        if (source is TimeAwareKObject<*>) {
            return "ModelEvent[src:[${source.now}]${source.path()}, type:$etype, elementAttributeType:$elementAttributeType, elementAttributeName:$elementAttributeName, value:$value, previousValue:$previous_value]";
        } else {
            return "ModelEvent[src:${source?.path()}, type:$etype, elementAttributeType:$elementAttributeType, elementAttributeName:$elementAttributeName, value:$value, previousValue:$previous_value]";
        }
    }

}
